### PR TITLE
Wait for 6 blocks after last transaction confirmation before writting another transaction

### PR DESCRIPTION
**Pull request by xinaxu**

### CODE REVIEW

File: lib/core/MongoDbConfirmationStore.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  public async confirm (anchorString: string, confirmedAt: number | undefined): Promise<void> {
+    await this.collection.findOneAndUpdate({ anchorString }, { confirmedAt });
+  }
+
+  public async getLastSubmitted (): Promise<{ submittedAt: number; confirmedAt: number | undefined } | null> {
+    const response: ConfirmationModel[] = await this.collection.find().sort({ submittedAt: -1 }).limit(1).toArray();
+    if (response.length === 0) {
+      return null;
```
Comment (thehenrytsai) [inline lib/core/MongoDbConfirmationStore.ts:27]: Use `undefined` as opposed to `null` if you can.

File: lib/core/versions/latest/BatchWriter.ts [hunk 2]
```diff
@@ -49,6 +52,14 @@ export default class BatchWriter implements IBatchWriter {
       return 0;
     }
 
+    const lastSubmitted = await this.confirmationStore.getLastSubmitted();
+    if (lastSubmitted !== null &&
+      (lastSubmitted.confirmedAt === undefined || currentTime.time - lastSubmitted.confirmedAt < this.minConfirmationBetweenWrites - 1)
```
Comment (thehenrytsai) [inline lib/core/versions/latest/BatchWriter.ts:57]: Why the -1 at the end?

File: lib/core/versions/latest/BatchWriter.ts [hunk 3]
```diff
@@ -101,6 +113,8 @@ export default class BatchWriter implements IBatchWriter {
     // Remove written operations from queue after batch writing has completed successfully.
     await this.operationQueue.dequeue(numberOfOperations);
 
+    await this.confirmationStore.submit(stringToWriteToBlockchain, currentTime.time);
```
Comment (thehenrytsai) [inline lib/core/versions/latest/BatchWriter.ts:116]: Do it before line 114, else might still write quicker than required.

File: lib/core/MongoDbConfirmationStore.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+interface ConfirmationModel {
+  anchorString: string;
+  submittedAt: number;
+  confirmedAt: number | undefined;
+}
+
+/**
+ * Implementation of LastWriteStore that stores the last update per write
+ */
+export default class MongoDbConfirmationStore extends MongoDbStore implements IConfirmationStore {
```
Comment (thehenrytsai) [inline lib/core/MongoDbConfirmationStore.ts:13]: Should we add index for both `submittedAt` and `anchorString`?

File: lib/core/Observer.ts [hunk 5]
```diff
@@ -124,6 +126,7 @@ export default class Observer {
           this.transactionsUnderProcessing.push(transactionUnderProcessing);
           // Intentionally not awaiting on downloading and processing each operation batch.
           this.processTransaction(transaction, transactionUnderProcessing);
+          await this.confirmationStore.confirm(transaction.anchorString, transaction.transactionTime);
```
Comment (thehenrytsai) [inline lib/core/Observer.ts:129]: Probably should do this within `processTransaction()` method given that `processTransaction()` is called in another place.

File: lib/core/Observer.ts [hunk 6]
```diff
@@ -340,6 +343,10 @@ export default class Observer {
 
     await this.unresolvableTransactionStore.removeUnresolvableTransactionsLaterThan(bestKnownValidRecentTransactionNumber);
 
+    for (const transaction of await this.transactionStore.getTransactionsLaterThan(bestKnownValidRecentTransactionNumber, undefined)) {
```
Comment (thehenrytsai) [inline lib/core/Observer.ts:346]: Can we do this by calling one method and pass in the transaction time? Something like `this.confirmationStore.removeConfirmationsLaterThan(bestKnownValidRecentTransactionTime)`?

File: lib/core/versions/latest/BatchWriter.ts (hunk 2 above)
Comment (xinaxu) [reply lib/core/versions/latest/BatchWriter.ts:57]: Say comfirmedAt = 100, currentTime = 105, then the message have already been confirmed 6 times in block 100 to 105. Hence it's okay to proceed. It only needs to wait if currentTime - confirmedAt < 5. What do you think?

File: lib/core/MongoDbConfirmationStore.ts (hunk 4 above)
Comment (xinaxu) [reply lib/core/MongoDbConfirmationStore.ts:13]: Yes

File: lib/core/Observer.ts (hunk 6 above)
Comment (xinaxu) [reply lib/core/Observer.ts:346]: That may be complicated. To do that, I'll need to add transaction time to the data model. And instead of removing the entry, I need to reset the confirmedAt field. Since the reorg will change the transactionTime, I will need to update the transactionTime as well.

File: lib/core/MongoDbConfirmationStore.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  constructor (serverUrl: string, databaseName: string) {
+    super(serverUrl, MongoDbConfirmationStore.collectionName, databaseName);
+  }
+
+  public async confirm (anchorString: string, confirmedAt: number): Promise<void> {
+    await this.collection.updateMany({ anchorString }, { $set: { confirmedAt } });
+  }
+
+  public async resetAfter (confirmedAt: number | undefined): Promise<void> {
+    await this.collection.updateMany({ confirmedAt: { $gt: confirmedAt } }, { $set: { confirmedAt: null } });
```
Comment (thehenrytsai) [inline lib/core/MongoDbConfirmationStore.ts:25]: Add a unit test for `undefined`, I think it's edge case, but since it can be passed, should clear the entire collection in which case.

File: lib/core/versions/latest/BatchWriter.ts [hunk 8]
```diff
@@ -27,11 +28,14 @@ import ValueTimeLockVerifier from './ValueTimeLockVerifier';
  * Implementation of the `IBatchWriter`.
  */
 export default class BatchWriter implements IBatchWriter {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/BatchWriter.ts:30]: For some reason branch coverage for BatchWriter is no longer 100%, bring it back please!

File: tests/json/config-test.json [hunk 9]
```diff
@@ -6,5 +6,6 @@
   "maxConcurrentDownloads": 20,
   "mongoDbConnectionString": "mongodb://localhost:27017/",
   "observingIntervalInSeconds": 10,
+  "minConfirmationBetweenWrites": 6,
```
Comment (thehenrytsai) [inline tests/json/config-test.json:9]: Remove.

File: lib/core/MongoDbConfirmationStore.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  public async getLastSubmitted (): Promise<{ submittedAt: number; confirmedAt: number | undefined } | undefined> {
```
Comment (thehenrytsai) [inline lib/core/MongoDbConfirmationStore.ts:28]: Consider return `ConfirmationModel` so you can log even more like what you've done with `submittedAt`.

**Comment by xinaxu** [comment]
All addressed.
@csuwildcat would you mind take a look and approve?

File: tests/core/BatchWriter.spec.ts [hunk 11]
```diff
@@ -1,29 +1,32 @@
...
 describe('BatchWriter', () => {
-  let blockchain: IBlockchain;
+  let blockchain: MockBlockchain;
   let cas: ICas;
   let operationQueue: IOperationQueue;
   let batchWriter: BatchWriter;
+  let confimrationStore: MockConfirmationStore;
 
   beforeAll(() => {
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000000;
```
Comment (thehenrytsai) [inline tests/core/BatchWriter.spec.ts:23]: What's the purpose of this?

**Review by thehenrytsai — APPROVED** [review]
